common/src/java/org/apache/hadoop/hive/common/FileUtils.java

```diff
@@ -493,12 +505,28 @@ private static void addChildren(FileSystem fsAsUser, Path path, List<FileStatus>
    */
   public static boolean isActionPermittedForFileHierarchy(FileSystem fs, FileStatus fileStatus,
                                                           String userName, FsAction action) throws Exception {
-    return isActionPermittedForFileHierarchy(fs,fileStatus,userName, action, true);
+    UserGroupInformation ugi = Utils.getUGI();
+    String currentUser = ugi.getShortUserName();
+    FileSystem fsAsUser = null;
+    UserGroupInformation proxyUser = null;
+    boolean isPermitted = false;
+    try {
+      if (userName != null && !userName.equals(currentUser)) {
+        proxyUser = UserGroupInformation.createProxyUser(userName, UserGroupInformation.getLoginUser());
+        fsAsUser = getFsAsUser(fs, proxyUser);
+      }
```

Comment:
Can you try moving this to separate method? This is duplicated in multiple places.

Reply:
Addressed